### PR TITLE
[IMP] stock: merge only quants related to actual unpack operation

### PR DESCRIPTION
**Pull request by grindtildeath**
Description of the issue/feature this PR addresses:

On a database with > 9 000 000 records in the stock_quant table, unpacking a Package (stock.quant.package) lasts more than 15-20 seconds.

Current behavior before PR:

The SQL query to merge quants targets the all the quants in the database.

Desired behavior after PR is merged:

The SQL query to merge quants targets only the quants that were modified by the unpack operation.

OPW-2984790

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_quant.py [hunk 1]
```diff
@@ -553,9 +554,20 @@ def _merge_quants(self):
...
                    DELETE FROM stock_quant WHERE id in (SELECT unnest(to_delete_quant_ids) from dupes)
         """
+        params = []
+        quant_tuples_to_merge = self.env.context.get("_quant_tuples_to_merge")
+        where_clause = ""
+        if quant_tuples_to_merge:
+            where_clause = """WHERE %s """ % """ OR """.join(
+                [
+                    """(product_id = %s AND location_id = %s AND company_id = %s)""" for quant_tuple in quant_tuples_to_merge
+                ]
```
Comment (gurneyalex) [inline addons/stock/models/stock_quant.py:564]:
```suggestion
                [
                    """(product_id = %s AND location_id = %s AND company_id = %s)""" 
                ] * len(quant_tuples_to_merge)
```

File: addons/stock/models/stock_quant.py [hunk 2]
```diff
@@ -553,9 +554,20 @@ def _merge_quants(self):
                         )
                    DELETE FROM stock_quant WHERE id in (SELECT unnest(to_delete_quant_ids) from dupes)
         """
+        params = []
+        quant_tuples_to_merge = self.env.context.get("_quant_tuples_to_merge")
```
Comment (gurneyalex) [inline addons/stock/models/stock_quant.py:558]: maybe add a comment stating where this context value comes from and what the expected format is.

**Review by StephaneMangin — APPROVED** [review]

**Comment by grindtildeath** [comment]
ping @adwid @amoyaux

**Comment by amoyaux** [comment]
Hello. I got the idea but it will add an extra search in smaller databases. Also for the performance issue I would like to have a profiler or an explain analyze of the problematic queries.

**Comment by grindtildeath** [comment]
Hello @amoyaux 
This is an explain plan from the prod DB before the change: https://explain.dalibo.com/plan/AYvJ
This is an explain plan from the same DB but on my local machine after the change: https://explain.dalibo.com/plan/LllI

**Comment by MiquelRForgeFlow** [comment]
@grindtildeath maybe you could do a backport of https://github.com/odoo/odoo/pull/121839 if possible, as that code seems more neat that yours 😉

**Comment by grindtildeath** [comment]
@MiquelRForgeFlow I won't change this PR now as this code is used in production for one of our biggest customers for a while and it works well.

Also I'm not sure which code is the more performant as:
 - I don't think the union of recordset for each pack being processed to select the quants is faster than using `mapped`, but I may be wrong
 - The SQL in this PR is more selective of the quants as the other one is using a combination of product_id and location_id, therefore not limited to the quants of packs being actually unpacked in the process.

Also I didn't know it's "allowed" to use `self._ids` in an `api.model` decorated method, but anyway, it's good that Odoo finally did something on a long standing issue.

**Comment by C3POdoo** [comment]
Dear @grindtildeath,

Thank you for your contribution but the version 14.0 is no longer supported.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._

**Pull request closed without merging**